integration_test/tests/free_and_cancel_tests.rs

```diff
@@ -0,0 +1,142 @@
+        let env_handle = allocate_env().unwrap();
+        let (conn_handle, stmt_handle) = connect_and_allocate_statement(env_handle, None);
+
+        (env_handle, conn_handle, stmt_handle)
+    }
+
+    /// This test is inspired by the SSIS Preview Data result set metadata flow.
+    /// It is altered to be more general than that specific flow, with a focus
+    /// on freeing the statement handle after gathering some metadata. This flow
+    /// depends on setup_connect_and_alloc_stmt.
+    /// After allocating a statement handle, the flow is:
+    ///     - SQLPrepareW(<query>)
+    ///     - SQLNumResultCols
+    ///     - <loop: for each column>
+    ///         - SQLDescribeColW
+    ///         - SQLColAttributeW(SQL_DESC_UNSIGNED)
+    ///         - SQLColAttributeW(SQL_DESC_UPDATABLE)
+    ///     - SQLFreeStmt(SQL_CLOSE)
+    ///     - SQLDisconnect
+    ///     - SQLFreeHandle(SQL_HANDLE_DBC)
+    ///     - SQLFreeHandle(SQL_HANDLE_ENV)
+    #[test]
+    fn test_free_stmt() {
+        let (env_handle, conn_handle, stmt_handle) = setup_connect_and_alloc_stmt();
+
+        unsafe {
+            let mut query: Vec<WideChar> =
+                cstr::to_widechar_vec("SELECT * FROM integration_test.foo");
+            query.push(0);
+            assert_eq!(
+                SqlReturn::SUCCESS,
+                SQLPrepareW(stmt_handle, query.as_ptr(), SQL_NTS as i32),
+            );
+
+            get_column_attributes(
+                stmt_handle as Handle,
+                2,
+                Some(vec![Desc::SQL_DESC_UNSIGNED, Desc::SQL_DESC_UPDATABLE]),
+                true,
+            );
+
+            assert_eq!(
+                SqlReturn::SUCCESS,
+                SQLFreeStmt(stmt_handle, FreeStmtOption::SQL_CLOSE)
+            );
+
+            disconnect_and_free_dbc_and_env_handles(env_handle, conn_handle);
+        }
+    }
+
+    /// This test is inspired by the SSIS Preview Data data retrieval flow.
+    /// It is altered to be more general than that specific flow, with a focus
+    /// on canceling the query after getting some data. This flow depends on
+    /// setup_connect_and_alloc_stmt.
+    /// After allocating a statement handle, the flow is:
+    ///     - SQLSetStmtAttrW(SQL_ATTR_QUERY_TIMEOUT, 15)
+    ///     - SQLExecDirectW(<query>)
+    ///     - SQLRowCount
+    ///     - <loop: for each column>
+    ///         - SQLColAttributeW(SQL_DESC_CONCISE_TYPE)
+    ///         - SQLColAttributeW(SQL_DESC_UNSIGNED)
+    ///         - SQLColAttributeW(SQL_COLUMN_NAME)
+    ///         - SQLColAttributeW(SQL_COLUMN_NULLABLE)
+    ///         - SQLColAttributeW(SQL_DESC_TYPE_NAME)
+    ///         - SQLColAttributeW(SQL_COLUMN_LENGTH)
+    ///         - SQLColAttributeW(SQL_COLUMN_SCALE)
+    ///     - <loop: until SQLFetch return SQL_NO_DATA>
+    ///         - SQLFetch
+    ///         - SQLGetData
+    ///     - SQLCancel
+    #[test]
+    fn test_cancel() {
```

Comment:
```suggestion
    fn test_cancel_noop() {
```
Aka, cancel will be a no-op in our case because the odbc version set for the environment is ODBC 3 and the query is not still executing.